integrade/tests/ui/utils.py

```diff
@@ -46,9 +46,14 @@ def get_element_depth(element):
 def find_element_by_text(driver, text, fail_hard=False):
     """Find an element which contains the given text."""
     def t(e):
-        return e.get_attribute('innerHTML')
+        try:
+            return e.__innerHTML
+        except AttributeError:
+            e.__innerHTML = e.get_attribute('innerHTML')
+            return e.__innerHTML
     elements = [
-        e for e in driver.find_elements_by_xpath('//*')
+        e for e in
+        driver.find_elements_by_xpath('//*[text()[contains(.,\'%s\')]]' % text)
```

Comment:
I know in discussions with @cdcabrera in the past, we have brought up the desirability of being able to find elements by ID instead of xpath. @elyezer might have more to say about that (I think he's had experience with xpath being fragile)

If there is something that would make your life easier to find elements, you should file an issue on frontigrade :)

Reply:
I wrote this utility specifically to avoid any dependency between our tests and the target project. Whether by ID or XPath, if we're relying on those internal details to locate elements I think our tests are more fragile. Instead, I wanted to present this as a way to target test elements in line with user perception. Either by finding an element by `#id_menu_user` or `//*/li[position()=2]`, our tests will be vulnerable to simple template or structure changes with no behavioural or functional difference.

Preferably we wouldn't put that cost on us or the frontend devs and incidental changes should be safe. Preferably we can test what the user sees, right? So, that's what I hoped to get with this utility and I'd like to reuse it to build tests that just say "See the element with the user's e-mail address on it (our menu label)? Click on that."

Reply:
That is a really good point. I like the approach on focusing more on what the user see instead of how the document internally is defined.